### ISSUE TITLE
BiFlowNet training

### ISSUE THREAD

**Issue by nguyendung1612** [ShanghaiTech-IMPACT/3D-MedDiffusion#17]
Hi, thank you for sharing this excellent work on 3D medical image generation.
I have two questions that I would like you to answer:

1) In the Implementation Details section, authors wrote the following: _"Regarding resolution, all competing methods, Ours (CT), and Ours (MR) were trained using a uniform resolution by resizing all images in the CTChestAbdomen dataset to 256× 256 × 256, while images in the MRBrain dataset were not resized, as they are originally at 192×192×192 size. However, for Ours (universal), uniform resolution was not applied."_

As I understand it, for **Ours (CT)** and **Ours (MR)**, you resampled the datasets into a single resolution for training. For **Ours (universal)**, you trained with multi-resolution. In your code, resolution is used as a condition input; I don't see how to train with multi-resolution with a batch size of 16.

2) In the Implementation Details section, authors wrote the following: _"We selected the best checkpoint to report both training time and evaluation metrics."_

In the file train_BiFlowNet_SingleRes.py, how can you determine the **best checkpoint** during training BiFlowNet?

Thank you for your time and assistance! Looking forward to your response.

**Comment by Mors20** [comment]
Thank you for your interest.

Regarding your first question, we do not currently provide the code for multi-resolution training. For reference, you can implement a resolution-aware sampler by grouping images with the same resolution into the same bucket and constructing each batch from a single bucket.

Regarding your second question, since unconditional generation methods typically do not have a standard validation set, we use two strategies to select the best checkpoint for each generative model:

1.We save checkpoints throughout training, run inference with each checkpoint, compute the evaluation metrics, and select the checkpoint with the best performance.
2.For diffusion-based methods, we additionally create a validation set and compute the L1 loss using the same timestep and noise realization for each checkpoint.

In our experiments, these two strategies generally select very similar, and often identical, checkpoints.

**Comment by nguyendung1612** [comment]
Thank you for your feedback and suggestions.

I used your pretrained model for inference. Your code works perfectly with the 4x pretrained model, but with the 8x pretrained model, the results are completely zero.
Is there a problem with the inference code or the 8x pretrained model?

Thank you for your time and assistance! Looking forward to your response.
